acados_template improvements

schwieni opened the pull request:
This PR adds some improvements to the exported code

1. help to identify constant time-steps with numerical differences (improves reliability when distinguishing code paths in template)
2. ocp & main: help reuse exported code, since the model is independent of the number of shooting nodes
3. sim & main: use dimension defines and help with code readability (use parameters instead of numbers in code)
4. example minimal_example_ocp_reuse_code.py added to ctest

FreyJo commented on interfaces/acados_template/acados_template/c_templates_tera/acados_solver.in.h:
```diff
@@ -144,6 +144,12 @@ nlp_solver_capsule * {{ model.name }}_acados_create_capsule(void);
 
 int {{ model.name }}_acados_create(nlp_solver_capsule * capsule);
+/**
+ * More generic acados_create function, since the number of stages usually is independent of the model and its
```
I suggest changing the comment to:
"Generic version of {{ model.name }}_acados_create which allows to use a different number of shooting intervals than the number used for code generation."

FreyJo commented on interfaces/acados_template/acados_template/acados_ocp_solver.py:
```diff
@@ -825,6 +831,10 @@ def __init__(self, acados_ocp, json_file='acados_ocp_nlp.json', simulink_opts=No
         assert getattr(self.shared_lib, f"{model.name}_acados_create")(self.capsule)==0
         self.solver_created = True
 
+        # create solver with another number of stages
```
I think these lines are not needed, since the solver in Python is still created via
the line above:
```
         assert getattr(self.shared_lib, f"{model.name}_acados_create")(self.capsule)==0
```

FreyJo commented on interfaces/acados_template/acados_template/c_templates_tera/acados_solver.in.c:
```diff
@@ -122,8 +122,23 @@ int {{ model.name }}_acados_free_capsule(nlp_solver_capsule *capsule)
+    int n_stages = {{ model.name | upper }}_N;
+    double* const_time_step = NULL; // NULL -> don't alter the exported value
+    return {{ model.name }}_acados_create_w_stages(capsule, n_stages, const_time_step);
+}
```
I suggest to change the lines above to:
```
    int N_shooting_intervals = {{ model.name | upper }}_N;
    double* const_time_step = NULL; // NULL -> use time step from code generation
    return {{ model.name }}_acados_create_with_discretization(capsule, N_shooting_intervals, const_time_step); 
```
I think the function name: `{{ model.name }}_acados_create_with_discretization` would better describe what it does.

FreyJo reviewed (changes requested):
Hi @schwieni ,

thanks for the PR!
I spotted one error and would like to rather use the term shooting intervals than stages. Just because number of stages is also used in the integrator settings.
In addition to the comments, I think it would be nicer to have an argument 
```
 int {{ model.name }}_acados_create_w_stages(nlp_solver_capsule * capsule, int N, double* time_steps)
```
where `time_steps` is a vector of time steps of length `N`.
What do you think?
Could you implement that and address the other comments?

Best,
Jonathan

schwieni commented:
Hi @FreyJo,
thank you for looking into this PR!
I will integrate your suggestions and make a commit.
Sorry about the (N+1)! Good that you are having a good eye on it. :-)

schwieni commented on interfaces/acados_template/acados_template/c_templates_tera/main.in.c:
```diff
@@ -42,12 +42,45 @@
+#define NR     {{ model.name | upper }}_NR
+
+
+int main(int argc, char** argv)
```
You are right. In a previous conversion we had, I understood that this file is considered a starting point for a c-based solver. So I was under the impression, that this would make sense. I will remove the signature again.

schwieni commented:
Hi @FreyJo,
I made the changes you requested above. Are you ok with the changes?
If you want I can also try to add an example in python also to test the feature (change N after compilation).
This could also be integrated in the ctests.

FreyJo commented on interfaces/acados_template/acados_template/c_templates_tera/main_sim.in.c:
```diff
@@ -41,6 +41,36 @@
+#define NHN    {{ model.name | upper }}_NHN
+#define NPHIN  {{ model.name | upper }}_NPHIN
+#define NR     {{ model.name | upper }}_NR
+
```
I would remove all dimensions but NX, NZ, NU and NP here, since this file is used to just test the standalone acados integrator.

FreyJo commented:
Thanks for addressing everything!
If you have such an example available and can add it to the `ctests` that would be nice.
I will wait for it and then merge this :+1:

FreyJo commented:
>  If you want I can also try to add an example in python also to test the feature (change N after compilation).

Do you still want to do this?
I can also merge it as is.

schwieni commented:
> > If you want I can also try to add an example in python also to test the feature (change N after compilation).
> 
> Do you still want to do this?
> I can also merge it as is.

Hi @FreyJo,
I added an example (minimal_example_ocp_reuse_code.py). I had to change some interfaces as well. I also added it to the list of ctest. See my last two commits.
I think now with the example and test the whole PR makes more sense. :-)
Can you review again, please?

FreyJo commented:
I would just "squash and merge" as-is, to keep the acados history linear.
Is that fine with you?

schwieni commented:
Hi @FreyJo,
> I really like the example.
cool :-)

> I would just "squash and merge" as-is, to keep the acados history linear.
> Is that fine with you?
Of course! Thanks for merging! :+1:
